### PR TITLE
GH-1158: Fix TopicPartitionInitialOff. Deprecation

### PR DESCRIPTION
**Pull request by garyrussell**
Revert "GH-1158: Deprecate TopicPartitionInitialOffset"

Previous commit caused a breaking API change.

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java [hunk 1]
```diff
@@ -379,13 +380,27 @@ else if (this.autoStartup != null) {
...
+	public C createContainer(final Collection<org.springframework.kafka.support.TopicPartitionInitialOffset> topicPartitions) {
+		return createContainer(topicPartitions.stream()
+				.map(TopicPartitionOffset::fromTPIO)
+				.collect(Collectors.toList())
+				.toArray(new TopicPartitionOffset[0]));
+	}
+
 	@Override
-	public C createContainer(final Collection<TopicPartitionOffset> topicPartitions) {
+	public C createContainer(final TopicPartitionOffset[] topicsAndPartitions) {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java]: Can this be a `varargs`?

File: spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java [hunk 2]
```diff
@@ -379,13 +380,27 @@ else if (this.autoStartup != null) {
...
 		KafkaListenerEndpoint endpoint = new KafkaListenerEndpointAdapter() {
 
 					@Override
-					public Collection<TopicPartitionOffset> getTopicPartitions() {
-						return topicPartitions;
+					public TopicPartitionOffset[] getTopicPartitionsToAssign() {
+						return topicsAndPartitions;
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java]: Don't we need to return a copy from here do not allow end-user to modify an internal state of this component?

File: spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java [hunk 3]
```diff
@@ -379,13 +380,27 @@ else if (this.autoStartup != null) {
...
+	 * {@inheritDoc}
+	 * @deprecated in favor of {@link #createContainer(TopicPartitionOffset[])}
+	 */
+	@Deprecated
+	@SuppressWarnings("deprecation")
+	@Override
+	public C createContainer(final Collection<org.springframework.kafka.support.TopicPartitionInitialOffset> topicPartitions) {
+		return createContainer(topicPartitions.stream()
+				.map(TopicPartitionOffset::fromTPIO)
+				.collect(Collectors.toList())
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java]: Doesn't `toArray(TopicPartitionOffset[]::new)` work here?

File: spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java [hunk 4]
```diff
@@ -379,13 +380,27 @@ else if (this.autoStartup != null) {
...
+	/**
+	 * {@inheritDoc}
+	 * @deprecated in favor of {@link #createContainer(TopicPartitionOffset[])}
+	 */
+	@Deprecated
+	@SuppressWarnings("deprecation")
+	@Override
+	public C createContainer(final Collection<org.springframework.kafka.support.TopicPartitionInitialOffset> topicPartitions) {
+		return createContainer(topicPartitions.stream()
+				.map(TopicPartitionOffset::fromTPIO)
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java]: Maybe such a mapper should be in an opposite place - `TopicPartitionInitialOffset.toTopicPartitionOffset()`?

Even if this is going to cause some class tangle, that mapper is going to disappear together with deleted deprecated class.

File: spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java [hunk 5]
```diff
@@ -190,6 +191,26 @@ public void setTopics(String... topics) {
...
+	 * @deprecated in favor of {@link #setTopicPartitions(TopicPartitionOffset...)}.
+	 * @see #setTopics(String...)
+	 * @see #setTopicPattern(Pattern)
+	 */
+	@Deprecated
+	@SuppressWarnings("deprecation")
+	public void setTopicPartitions(org.springframework.kafka.support.TopicPartitionInitialOffset... topicPartitions) {
+		Assert.notNull(topicPartitions, "'topics' must not be null");
+		this.topicPartitions.clear();
+		this.topicPartitions.addAll(Arrays.asList(topicPartitions).stream()
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java]: `Arrays.stream(topicPartitions)`?

File: spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java [hunk 6]
```diff
@@ -190,6 +191,26 @@ public void setTopics(String... topics) {
...
+				.map(TopicPartitionOffset::fromTPIO)
+				.collect(Collectors.toList()));
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java]: `forEach(this.topicPartitions::add)` ?

File: spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java [hunk 7]
```diff
@@ -202,10 +223,25 @@ public void setTopicPartitions(TopicPartitionOffset... topicPartitions) {
...
 	 * Return the topicPartitions for this endpoint.
 	 * @return the topicPartitions for this endpoint.
+	 * @deprecated in favor of {@link #getTopicPartitionsToAssign()}
+	 */
+	@Deprecated
+	@SuppressWarnings("deprecation")
+	@Override
+	public Collection<org.springframework.kafka.support.TopicPartitionInitialOffset> getTopicPartitions() {
+		return Collections.unmodifiableCollection(this.topicPartitions.stream()
+				.map(TopicPartitionOffset::toTPIO)
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerEndpoint.java]: Same "ball of mud" isolation concern: `TopicPartitionInitialOffset.fromTopicPartitionOffset()`.

File: spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java (hunk 1 above)
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java]: No; `@SafeVarArgs` can't be used on a non-final method.

Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java]: My plan is to deprecate this in 2.4 and reintroduce `Collection<TPO>`.

Comment (artembilan) [reply spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java]: Why do we talk about `@SafeVarArgs` here if our type is just an explicit and there is no any generics?
😕

Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/config/AbstractKafkaListenerContainerFactory.java]: Not sure what happened, but I got a warning yesterday; now it's not complaining - fixing.

**Comment by artembilan** [comment]
Gary, here is some polishing for your confirmation: https://github.com/artembilan/spring-kafka/commit/e7374d3107c16dbac965f1545bd10a2afc131025

Pay attention to the redundant suppresses when we already have a `@Deprecated` and also to the `default` method in the `KafkaListenerEndpoint`.
Plus some goodies from Java Streams.

Everything else is like inconsistency between our IDEs. 😢

**Comment by garyrussell** [comment]
LGTM; thanks.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/2a05612410788ddaa1527da2ad9e4dd0ed63486e.

**Pull request closed without merging**